src/ucs/sys/topo.c

```diff
@@ -137,15 +137,12 @@ ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
      * TODO implement more accurate estimation, based on system type, PCIe
      * switch, etc.
      */
-    if (path_distance <= 2) {
+    if (path_similarity <= 2) {
         distance->latency   = 0;
         distance->bandwidth = DBL_MAX;
-    } else if (path_distance <= 4) {
-        distance->latency   = 300e-9;
-        distance->bandwidth = 2000 * UCS_MBYTE;
     } else {
-        distance->latency   = 900e-9;
-        distance->bandwidth = 300 * UCS_MBYTE;
+        distance->latency   = 900e-9 / path_similarity;
+        distance->bandwidth = 300 * path_similarity * UCS_MBYTE;
```

Comment:
ok, but we should calc based on "distance" and not similarity
paths /a/b/c/d-/a/b/c/x should exhibit same bw/lat as b/c/d-b/c/x

Reply:
Just to make sure I understand the above comment correctly - 
ucs_topo_get_distance should provide the right relative distance w.r.t a given a device. For any x, all devices of the form /a/b/c/x are equidistant from /a/b/c/d. Let's say that bw/lat = (p, q). The above code change should make sure that this is true. We shouldn't be comparing (p,q) with other devices /b/c/y that equidistant from /b/c/d. If the latter bw/lat = (m, n), the expectation shouldn't be that (m, n) = (p, q) right? If I'm not mistaken, the absolute distance is not needed by UCP. For example, we're interested in comparing distance(NIC0, GPU0) and distance(NIC1, GPU0); but maybe not comparing distance(NIC1, GPU1) and distance(NIC0, GPU0).

Reply:
ok, now that it changed to "distance" the calculation does not make sense to me: for larger distance, latency should grow and bandwidth should drop, not the other way around

Reply:
That's the reason I had changed the name from "distance" to "similarity". With greater similarity the lower the latency and greater the bandwidth and vice versa.

Reply:
in the example above, IMO the bw,lat of [/a/b/c/x , /a/b/c/d] should be same as [/b/c/y , /b/c/d]
however, in the "similarity" code, the bw,lat of the former would be better

Reply:
> IMO the bw,lat of [/a/b/c/x , /a/b/c/d] should be same as [/b/c/y , /b/c/d]

@yosefe To satisfy this, seems like we need to look at common ancestor of two paths. There's always one common ancestor `/` for all paths and so we will always have some valid path distance between the two given devices. How should we handle the path distance between two given paths? Should the distance be min/max/sum from common ancestor? (in which case distance would be 1/2/3 between "/a/b/c/d" and "/a/b/c/e/f"). Sum would be similar to counting the hops between the two devices but there can be some issues. For the following example,

```
a
├── b
│   ├── c
│   │   └── f
│   └── d
│       └── g
└── e
```

hops(f,g) = 3 (f->c->b->d->g) and hops(f,e) = 3 (f->c->b->a->e)
Even though # hops is the same, path(f,g) would have better latency/bandwidth. We could do realized that with weighted hop count. Any other corner cases we could be missing from using weighted hop count? or other thoughts?

Reply:
IMO, without using hwloc or topo xml, it would never be as good as we need.. so we could use max() of distances for now, should cover both cases, WDYT?

Reply:
@yosefe altered UCS path_distance and topo distance fxn based on the discussion above.